languages/python/reference/exercise-concepts/robot-simulator.md

```diff
@@ -0,0 +1,62 @@
+
+- classes: The exercise objective is to define a `robot` type.  Tested methods are linked to a `robot` class.
+
+- objects: Creating different instances of the `robot` class with different data representing different starting positions and bearing are tested.
+
+- constructor: Customizing object initalization with actions and persisting data. The example uses a constructor (via `__init__` + `__new__`) to persist a `compass` object and x, y coordinates assigned to instance properties.
+
+- retun values:  Knowing that functions need not have explicit return statements or values but will return `none` if `return` is not specified.  Except for the two `@property` functions, all of the functions in the example omit an explicit `return` statments, and all return `none`.
+
+- self: The example uses this keyword for methods and properties linked to a specific instance of the class.
+
+- namespaces: Knowing to use self.<propertyname> for instance properties and self as first argument to instance methods in a class.  Additionally, the example uses `self.<methodname>()` to call a previously stored function name.
+  
+- instance methods: Tests for this exercises require one or more instance methods that will take in a set of starting coordinates and a bearing and then accept a seiries of instructions that "move" the instance to a new set of coordinates and bearing.
+
+- function decorator: a higer-order function that takes another function as an argument. The "decorating" function extends the behavior of the "decorated" function without explicitly modifying it (i.e. it _wraps_ or _decorates_ it). Called in python by using the `@` symbol and the function name ahead of the function being decorated.  The example uses pythons built-in `property()` as a decorator (`@property`) to return a "compound" read-only instance property made up of two seperate instance properties.
+
+- higher-order functions:  a function that takes one or more other functions as arguments, _returning_ a function as its retrun value.  The example uses the built-in `property()` as a higher-order function through `@property`.
+
+- `property()` built-in:  a function that returns a property attribute.  When used as a decorator, this transforms the passed-in method into a _getter_ method for read-only attribute with the same name and docstring.
+
+- assignment:  the example uses assignment for all the instance properties and `instructions` dictionary.
+
+- instance properties: This exercise rquires one or more instance properties to persist passed in data.
+
+- mutability: In the example, knowing there are no protected or private properties in python and so conciously mutating `self.x`, `self.y` and `self.compass` through the called instance methods.
+
+- method arguments: the example `__init__` method needs to take `self`, direction, x, and y (coordinates) as arguments.  It also uses `self` and `commands` (a string) for arguments to the `move()` method.
```

Comment:
It's a bit of a pedantic distinction, but the function signature declares _parameters_ the function / method requires, _arguments_ are more accurately the values passed in when the function / method is actually called. For instance `*args* and `**kwargs` are both variadic parameters that may never receive arguments.

Reply:
Rephrased this, although it feels a bit awkward.  Probably because I've been mis-using _arguments_  ;-):

> - method parameters: the example `__init__` method has `self`, direction, x, and y (coordinates) as parameters.  It also uses `self` and `commands` (a string) for parameters of the `move()` method.

Reply:
I think the new wording is okay, though maybe add a line to clarify, along the lines of "A user-provided value assigned to a given parameter is called an argument."